### PR TITLE
Splitter:  add ResizeHandle component

### PR DESCRIPTION
**Pull request by AlexanderMoiseev**

### CODE REVIEW

File: packages/devextreme/js/__internal/ui/splitter/resize_handle.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+  _toggleDirection(): void {
+    this.$element().toggleClass(HORIZONTAL_DIRECTION_CLASS, this._isHorizontalDirection());
+    this.$element().toggleClass(VERTICAL_DIRECTION_CLASS, !this._isHorizontalDirection());
+  }
+
+  _render(): void {
+    super._render();
+
+    this._detachEventHandlers();
```
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/splitter/resize_handle.ts]: I think we shouldn't detach events here

File: packages/devextreme/js/__internal/ui/splitter/resize_handle.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  }
+
+  _initMarkup(): void {
+    super._initMarkup();
+
+    this._toggleDirection();
+    this.$element().addClass(RESIZE_HANDLE_CLASS);
+  }
+
+  _toggleDirection(): void {
```
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/splitter/resize_handle.ts]: It's better to name this method '_toggleDirectionClass'

File: packages/devextreme/js/__internal/ui/splitter/resize_handle.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    switch (args.name) {
+      case 'direction':
+        this._toggleDirection();
+        this._detachEventHandlers();
+        this._attachEventHandlers();
+        break;
+      case 'onResize':
+      case 'onResizeStart':
+      case 'onResizeEnd':
+        this._renderActions();
```
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/splitter/resize_handle.ts]: Should we recreate all actions when one of these callbacks was changed? I think no

File: packages/devextreme/js/__internal/ui/splitter/resize_handle.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  }
+
+  _resizeEndHandler(e: ResizeEndEvent): void {
+    this._resizeEndAction({
+      event: e,
+    });
+  }
+
+  _renderActions(): void {
+    this._resizeStartAction = this._createActionByOption('onResizeStart');
```
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/splitter/resize_handle.ts:85]: Should we add exclude validators for all events?
```
{
       excludeValidators: ['readOnly'] // disabled
 }
```
You may give answer later. Please extract the separate subtask.

File: packages/devextreme/js/__internal/ui/splitter/resize_handle.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    });
+  }
+
+  _detachEventHandlers(): void {
+    // @ts-expect-error todo: make optional parameters for eventsEngine
+    eventsEngine.off(this.$element(), this.handlers);
+  }
+
+  _isHorizontalDirection(): boolean {
+    return this.option('direction') === 'horizontal';
```
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/splitter/resize_handle.ts]: Can we extract 'horizontal', 'vertical' values to separate object 'DIRECTION' and use it instead as follows

https://github.com/EugeniyKiyashko/DevExtreme/blob/5296f968dc85e0004ac3e961feb03ddeaa7c11e7/packages/devextreme/js/ui/tabs.js#L98-L101

File: packages/devextreme/scss/widgets/base/splitter/_index.scss [hunk 6]
```diff
@@ -10,3 +11,9 @@
     flex-direction: column;
   }
 }
+
+.dx-resize-handle {
```
Comment (EugeniyKiyashko) [inline packages/devextreme/scss/widgets/base/splitter/_index.scss]: If you'd like to add this style, let's specify the details for both scenarios

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.markup.tests.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    QUnit.test('should have correct root class', function(assert) {
+        assert.strictEqual(this.instance.$element().hasClass(RESIZE_HANDLE_CLASS), true);
+    });
+
+    QUnit.test('should be initialized with horizontal class by default', function(assert) {
+        assert.strictEqual(this.instance.$element().hasClass(VERTICAL_DIRECTION_CLASS), false);
+        assert.strictEqual(this.instance.$element().hasClass(HORIZONTAL_DIRECTION_CLASS), true);
+    });
+
+    QUnit.test('direction should be initialized correctly', function(assert) {
```
Comment (EugeniyKiyashko) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.markup.tests.js:51]:
```suggestion
    QUnit.test('should be initialized with vertical class if direction is vertical', function(assert) {
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.markup.tests.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        this.reinit({ direction: 'vertical' });
+
+        assert.strictEqual(this.instance.$element().hasClass(HORIZONTAL_DIRECTION_CLASS), false);
+        assert.strictEqual(this.instance.$element().hasClass(VERTICAL_DIRECTION_CLASS), true);
+    });
+
+    QUnit.test('direction should be changed at runtime', function(assert) {
```
Comment (EugeniyKiyashko) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.markup.tests.js:58]:
```suggestion
    QUnit.test('direction class should be changed correctly after change direction option at runtime', function(assert) {
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.tests.js [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+            init(options);
+        };
+    },
+    afterEach: function() {
+        fx.off = false;
+    }
+};
+
+QUnit.module('Initialization', moduleConfig, () => {
+    QUnit.test('ResizeHandle should be initialized with correct type', function(assert) {
```
Comment (EugeniyKiyashko) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.tests.js]: I think we should change the test name

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.tests.js [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+        assert.ok(this.instance instanceof ResizeHandle);
+    });
+
+    QUnit.test('ResizeHandle  should be initialized with horizontal direction by default', function(assert) {
```
Comment (EugeniyKiyashko) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.tests.js]: Why do you need these test? We check this using defaultOptions.tests.

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.tests.js [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+        assert.strictEqual(this.instance.option('direction'), 'horizontal');
+    });
+
+    QUnit.test('every resizeHandle should be initialized with unique name', function(assert) {
```
Comment (EugeniyKiyashko) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.tests.js]:
```suggestion
    QUnit.test('resizeHandle events should be created with unique names after initialization', function(assert) {
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.tests.js [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+        const pointer = pointerMock(this.instance.$element());
+
+        pointer.start().dragStart().drag(0, 50).drag(0, 50).dragEnd();
+
+        assert.ok(onResizeStartStub.calledOnce);
+        assert.ok(onResizeStub.calledTwice);
+        assert.ok(onResizeEndStub.calledOnce);
+    });
+
+    QUnit.test('resize event handlers should be able to update at runtime', function(assert) {
```
Comment (EugeniyKiyashko) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.tests.js]: We should test single scenario in test.

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.tests.js [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+    QUnit.test('resize event handlers should be correctly added with "on" function', function(assert) {
+        const onResizeStartStub = sinon.stub();
+        const onResizeStub = sinon.stub();
+        const onResizeEndStub = sinon.stub();
+
+        this.instance.on('resize', onResizeStub);
+        this.instance.on('resizeStart', onResizeStartStub);
+        this.instance.on('resizeEnd', onResizeEndStub);
+
+        const pointer = pointerMock(this.instance.$element());
```
Comment (EugeniyKiyashko) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.tests.js]: You can declare this.$element = this.instance.$element() in module config

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.tests.js [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+        assert.ok(newOnResizeEndStub.calledOnce);
+
+        assert.ok(onResizeStartStub.calledOnce);
+        assert.ok(onResizeStub.calledTwice);
+        assert.ok(onResizeEndStub.calledOnce);
+    });
+
+    QUnit.test('event handlers should recieve correct arguments', function(assert) {
+        const checkArgs = (e) => {
+            assert.ok(e.event);
```
Comment (EugeniyKiyashko) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.tests.js]: These asserts is to simple. We want have something like this https://github.com/EugeniyKiyashko/DevExtreme/blob/5296f968dc85e0004ac3e961feb03ddeaa7c11e7/packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/dateRangeBox.tests.js#L1532

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.tests.js [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+                checkArgs(e);
+            },
+        });
+
+        const pointer = pointerMock(this.instance.$element());
+
+        pointer.start().dragStart().drag(0, 10).dragEnd();
+    });
+
+    QUnit.test('events should be called once after direction option changed', function(assert) {
```
Comment (EugeniyKiyashko) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/resizeHandle.tests.js:138]: Why we check only resizeStart event?

File: packages/devextreme/scss/widgets/base/splitter/_index.scss (hunk 6 above)
Comment (AlexanderMoiseev) [reply packages/devextreme/scss/widgets/base/splitter/_index.scss]: i'll remove it for now

File: packages/devextreme/js/__internal/ui/splitter/resize_handle.ts [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+  _renderResizeStartAction(): void {
+    this._resizeStartAction = this._createActionByOption('onResizeStart');
+  }
+
+  _renderResizeEndAction(): void {
+    this._resizeEndAction = this._createActionByOption('onResizeEnd');
+  }
+
+  _attachEventHandlers(): void {
+    this.handlers[this.RESIZE_START_EVENT_NAME] = this._resizeStartHandler.bind(this);
```
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/splitter/resize_handle.ts:94]: this.resizeHandlers = {
    [`this.RESIZE_START_EVENT_NAME`]: this._resizeStartHandler.bind(this),
    ...
}

File: packages/devextreme/js/__internal/ui/splitter/resize_handle.ts [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+    this.handlers[this.RESIZE_EVENT_NAME] = this._resizeHandler.bind(this);
+    this.handlers[this.RESIZE_END_EVENT_NAME] = this._resizeEndHandler.bind(this);
+
+    eventsEngine.on(this.$element(), this.handlers, {
```
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/splitter/resize_handle.ts]: Let's extract event data to separate variable and declare three .on function calls.

File: packages/devextreme/js/__internal/ui/splitter/resize_handle.ts [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+    });
+  }
+
+  _init(): void {
+    super._init();
+    const namespace = `${RESIZE_HANDLER_MODULE_NAMESPACE}${new Guid().toString()}`;
+    this.RESIZE_START_EVENT_NAME = addNamespace(dragEventStart, namespace);
+    this.RESIZE_EVENT_NAME = addNamespace(dragEventMove, namespace);
+    this.RESIZE_END_EVENT_NAME = addNamespace(dragEventEnd, namespace);
+    this.handlers = {};
```
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/splitter/resize_handle.ts]: Let's remove this.handlers

File: packages/devextreme/js/__internal/ui/splitter/resize_handle.ts [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+    });
+  }
+
+  _resizeEndHandler(e: ResizeEndEvent): void {
+    this._resizeEndAction({
+      event: e,
+    });
+  }
+
+  _renderResizeAction(): void {
```
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/splitter/resize_handle.ts]:
```suggestion
  _createResizeAction(): void {
```

**Review by EugeniyKiyashko — APPROVED** [review]

**Pull request merged**